scipy/_lib/_array_api.py

```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
     else:
         # to maintain backwards compatibility
         return np.linalg.norm(x, ord=ord, axis=axis, keepdims=keepdims)
+
+
+# work around xp.real raising for floats etc
+def xp_real(a, xp=None):
+    xp = array_namespace(a) if xp is None else xp
+    if xp.isdtype(a.dtype, 'complex floating'):
+        return xp.real(a)
+    else:
+        return a
+
+
+# work around xp.real raising for floats etc
+def xp_imag(a, xp=None):
+    xp = array_namespace(a) if xp is None else xp
+    if xp.isdtype(a.dtype, 'complex floating'):
+        return xp.imag(a)
+    else:
+        return xp.zeros_like(a)
+
+
+# work around xp.real raising for floats etc
+def xp_conj(a, xp=None):
+    xp = array_namespace(a) if xp is None else xp
+    if xp.isdtype(a.dtype, 'complex floating'):
+        return xp.conj(a)
+    else:
+        return a
+
+
+def xp_ravel(a, xp):
+    return xp.reshape(a, (-1,))
+
+
+def xp_isscalar(a, xp):
+    return xp.asarray(a).ndim == 0
```

Comment:
This is incorrect, e.g.:
```python
>>> def xp_isscalar(a, xp=np):
...     return xp.asarray(a).ndim == 0
...
>>> np.isscalar(None)
False
>>> xp_isscalar(None)
True
```

`np.isscalar` is not a great function. Code that uses it now should usually use `a.ndim == 0`, which works for all array types and for numpy scalars. Please don't add it here.